allenact/embodiedai/models/visual_nav_models.py

```diff
@@ -0,0 +1,210 @@
+
+        self.belief_names = list(self.state_encoders.keys())
+
+        get_logger().info(
+            "there are {} belief models: {}".format(
+                len(self.belief_names), self.belief_names
+            )
+        )
+
+    def create_actorcritic_head(self):
+        self.actor = LinearActorHead(self._hidden_size, self.action_space.n)
+        self.critic = LinearCriticHead(self._hidden_size)
+
+    def create_aux_models(self, obs_embed_size: int, action_embed_size: int):
+        if self.auxiliary_uuids is None:
+            return
+        aux_models = OrderedDict()
+        for aux_uuid in self.auxiliary_uuids:
+            aux_models[aux_uuid] = AuxiliaryModel(
+                aux_uuid=aux_uuid,
+                action_dim=self.action_space.n,
+                obs_embed_dim=obs_embed_size,
+                belief_dim=self._hidden_size,
+                action_embed_size=action_embed_size,
+            )
+
+        self.aux_models = nn.ModuleDict(aux_models)
+
+    @property
+    def num_recurrent_layers(self):
+        """Number of recurrent hidden layers."""
+        return list(self.state_encoders.values())[0].num_recurrent_layers
+
+    @property
+    def recurrent_hidden_state_size(self):
+        """The recurrent hidden state size of a single model.
+        """
+        return self._hidden_size
+
+    def _recurrent_memory_specification(self):
+        return {
+            memory_key: (
+                (
+                    ("layer", self.num_recurrent_layers),
+                    ("sampler", None),
+                    ("hidden", self.recurrent_hidden_state_size),
+                ),
+                torch.float32,
+            )
+            for memory_key in self.belief_names
+        }
+
+    def forward_encoder(self, observations: ObservationType) -> torch.FloatTensor:
+        raise NotImplementedError("Obs Encoder Not Implemented")
+
+    def fuse_beliefs(
+        self, beliefs_dict: Dict[str, torch.FloatTensor], obs_embeds: torch.FloatTensor,
+    ) -> Tuple[torch.FloatTensor, Optional[torch.FloatTensor]]:
+        total_beliefs = torch.stack(list(beliefs_dict.values()), dim=-1)  # (T, N, H, k)
+
+        if self.multiple_beliefs:  # call the fusion model
+            return self.fusion_model(total_beliefs=total_beliefs, obs_embeds=obs_embeds)
+        # single belief
+        beliefs = total_beliefs.squeeze(-1)  # (T,N,H)
+        return beliefs, None
+
+    def forward(  # type:ignore
+        self,
+        observations: ObservationType,
+        memory: Memory,
+        prev_actions: torch.Tensor,
+        masks: torch.FloatTensor,
+    ) -> Tuple[ActorCriticOutput[DistributionType], Optional[Memory]]:
+
+        # 1.1 use perception model (i.e. encoder) to get observation embeddings
+        obs_embeds = self.forward_encoder(observations)
+        # 1.2 use embedding model to get prev_action embeddings
+        prev_actions_embeds = self.prev_action_embedder(prev_actions).to(obs_embeds)
```

Comment:
Is the `.to(obs_embeds)` necessary here? Shouldn't `prev_actions` already be on the appropriate gpu?

Reply:
Not really. In `FeatureEmbedding` class, its forward function may return a `torch.zeros(0,)` because it did not know the device when created.

Reply:
Got it. The way to make this work then is to make `torch.zeros(0,)` a buffer inside `FeatureEmbedding`:
```python
class FeatureEmbedding(nn.Module):
  def __init__(...):
    ...
    self.null_embedding: torch.Tensor
    self.register_buffer("null_embedding", torch.zeros(0,), persistent=False)
    ...

  def forward(self, inputs):
    if self.output_size != 0:
      return self.fc(inputs)
    else:  # useful for concat, to be move to device
      return self.null_embedding
```

Reply:
Interesting trick